rfcs/20210119-determinism.md

```diff
@@ -0,0 +1,113 @@
+# RFC: Enabling Determinism in TensorFlow
+  
+| Status        | Proposed                                                                    |
+:---------------|:----------------------------------------------------------------------------|
+| **Author(s)** | Pankaj Kanwar (Google),Reed Wanderman-Milne (Google), Duncan Riach (NVIDIA) |
+| **Sponsor**   | Sanjoy Das (Google)                                                         |
+| **Updated**   | 2021-01-31                                                                  |
+
+## Objective
+Allow users to enable deterministic behavior in TensorFlow. This means if the user runs a TensorFlow program multiple times, the model outputs and weights will be the same each time. Determinism will be supported on CPUs and GPUs.
+
+To get deterministic behavior, users must do the following:
+
+* Enable determinism using the API proposed in this doc.
+* Use same hardware configuration in every run.
+* Use the same software environment every run (OS, checkpoints, version of CUDA and TF, environmental variables, etc).
+* Not use constructs outside TensorFlow that are nondeterministic, such as Python’s `random` module (without a fixed seed) or using multiple threads/processes in ways that influence TensorFlow’s behavior.
+* Not use nondeterministic custom ops.
+
+## Motivation
+There are several mission critical applications in medicine, finance and automation that require deterministic behavior. Determinism is required so that the behavior of these applications can be accurately predicted & demonstrated in a variety of scenarios.
+
+Lack of determinism prevents companies from launching products using models developed in TF. For a subset of these industries having deterministic behavior is a regulatory requirement. 
+
+In addition, determinism increases model velocity development by reducing noise, while also simplifying the debugging workflow.
+
+## Design Proposal
+We will create a new flag with the default value of ‘False’ which enables determinism.  We will define 2 functions:
+
+* `tf.config.enable_deterministic_execution(enabled)`
+* `tf.config.deterministic_execution_enabled()`
+
+The first function takes in a boolean value, and allows the model developer to enable/disable determinism. The second function returns a bool indicating whether determinism is enabled.
+In some cases, we have deterministic and nondeterministic versions of the kernel. In such cases, we will use this flag to run the appropriate kernels.
+For ops which do not yet have a deterministic implementation, TensorFlow will raise a `tf.errors.UnimplementedError` if the flag is enabled.
```

Comment:
I wonder if it's worth clarifying here:

When deterministic execution is enabled, `tf.errors.UnimplementedError` will be thrown if a nondeterministic code path through an op would otherwise be traversed (either in eager mode or during graph execution). Note that it may still be possible to simply construct a graph containing ops that have nondeterministic code paths through them without the error being thrown.

Reply:
I added the sentences

> Certain ops will only raise an error for certain input shapes or attributes. Depending on the op, in graph mode, the error will either be raised when the op is constructed or when the op is run.

The first sentence addresses your first point, as it implies that only certain nondeterministic codepaths raise errors. For you second point: it depends on whether the op raises an error when it is constructed or run. I think both are acceptable, so in the RFC I left this up to each op's author.